### PR TITLE
fix: masking on macos

### PR DESCRIPTION
**Pull request by ajbt200128**
MacOS ridiculously symlinks their `$TMPDIR =
/var/folders/lz/<somestring>/T/`
to `/private/var/folders/...`. This means that if for whatever reason someone prints the real path of a tmp folder instead of the symlinked path, TestO wouldnt mask it.

Also, $TMPDIR on MacOS has a trailing slash for some reason, when on Linux `Filename.get_temp_dir_name` does not which also broke some things.

Note that I didn't test this on linux, and am hoping there are some CI tests there, but if not will need someone to test it for me

### CODE REVIEW

**Comment by ajbt200128** [comment]
Apologies for the extra diffs, my ocaml formatter does that for some reason in this repo

File: core/Testo.ml [hunk 1]
```diff
@@ -191,79 +192,78 @@ let mask_line ?(mask = "<MASKED>") ?(after = "") ?(before = "") () =
...
+let mask_temp_paths ?(depth = Some 1) ?replace
+    ?(tmpdir = Filename.get_temp_dir_name ()) () =
+  (* Some OS's *cough cough MacOS* symlink their $TMPDIR to a different path *)
+  (* This means that if an ocmal program ever uses realpath, they may print a valid *)
+  (* But different temp dir that's not equal to $TMPDIR *)
+  (* So let's include that in the pattern *)
+  let real_tmpdir = Unix.realpath tmpdir in
+  let tmpdir =
+    if String.ends_with ~suffix:Filename.dir_sep tmpdir then
+      String.sub tmpdir 0 (String.length tmpdir - 1)
```
Comment (kopecs) [inline core/Testo.ml:245]:
```suggestion
      String.sub tmpdir 0 (String.length tmpdir - String.length Filename.dir_sep)
```

Comment (kopecs) [reply core/Testo.ml:245]: Probably slightly more communicative

File: core/Testo.ml [hunk 2]
```diff
@@ -191,79 +192,78 @@ let mask_line ?(mask = "<MASKED>") ?(after = "") ?(before = "") () =
...
-    ?(depth = Some 1)
-    ?replace
-    ?(tmpdir = Filename.get_temp_dir_name ())
-    () =
+let mask_temp_paths ?(depth = Some 1) ?replace
+    ?(tmpdir = Filename.get_temp_dir_name ()) () =
+  (* Some OS's *cough cough MacOS* symlink their $TMPDIR to a different path *)
+  (* This means that if an ocmal program ever uses realpath, they may print a valid *)
+  (* But different temp dir that's not equal to $TMPDIR *)
+  (* So let's include that in the pattern *)
```
Comment (kopecs) [inline core/Testo.ml:241]: Maybe just 1 comment?

**Review by kopecs — COMMENTED** [review]
Are these tests in this repo? Would probably best to add some for this if we have them set up.

File: core/Testo.ml [hunk 3]
```diff
@@ -191,79 +192,78 @@ let mask_line ?(mask = "<MASKED>") ?(after = "") ?(before = "") () =
...
+  let real_tmpdir = Unix.realpath tmpdir in
+  let tmpdir =
+    if String.ends_with ~suffix:Filename.dir_sep tmpdir then
+      String.sub tmpdir 0 (String.length tmpdir - 1)
+    else tmpdir
+  in
   let tmpdir_pat = Re.Pcre.quote tmpdir in
+  let real_tmpdir_pat = Re.Pcre.quote real_tmpdir in
+  let tmpdir_pat = Printf.sprintf "(?:%s|%s)" tmpdir_pat real_tmpdir_pat in
```
Comment (mjambon) [inline core/Testo.ml:250]: Looks good to me. One issue is that `Unix.realpath` requires OCaml >= 4.13 which is now 3 years old. It's fine if we really have to but not ideal. `String.ends_with` also requires OCaml >= 4.13 but is easier to reimplement ourselves.

If I understand correctly, there are two issues:
1. The trailing slash
2. Multiple temp folders

For (1), we need tests. In particular, we need to check that `"/tmp//"` works.

For (2), we also need tests. If depending on OCaml 4.13 is an issue, we could leave this responsibility to the user by letting them specify multiple temp folders. They can actually already do this today by applying two filters: one with the original path obtained with `Filename.get_temp_dir_name ()` and another that is `Filename.get_temp_dir_name () |> Unix.realpath`.

Comment (mjambon) [reply core/Testo.ml:250]: Another issue: `String.sub tmpdir 0 (String.length tmpdir - 1)` assumes the length of `Filename.dir_sep` is 1 which is not guaranteed. Better use `String.length tmpdir - String.length Filename.dir_sep`.

**Comment by mjambon** [comment]
I redid this for compatibility with OCaml 4.08 and left the call to `Unix.realpath` to be the user's responsibility (see https://github.com/semgrep/semgrep/pull/10018) since it requires OCaml >= 4.13 and it's hard to reimplement.

**Pull request closed without merging**